### ISSUE TITLE
Tester races on timeout (casper.start doesn't clear old intervals)

### ISSUE THREAD

**Issue by james-w** [casperjs/casperjs#1766]
Hi,

We've found a problem where the tester has a race condition, but I think the
problem is a little more general than the tester.

tl;dr the tester races as casper.start doesn't clear the active intervals, so
an wait from a previous test can fail a later one

You can reproduce this with these two test files:

`niceTest.js`:

    'use strict';

    casper.test.begin('Nice test', function(test) {
      casper.start('http://google.com');
      casper.wait(1500);
      casper.then(function () { test.assert(true); });
      casper.run(function() { test.done(); });
    });

`nastyTest.js`:

    'use strict';

    casper.test.begin('Nasty test', function(test) {
      casper.start('http://google.com');

      casper.then(function() {
        test.assertVisible('.nonExistingClass', 'FORCING ERROR');
      });

      casper.waitForSelector('.notExisting0', function() {
        test.assertExists('.notExisting0', 'THIS SHOULD NOT HAPPEN');
        this.captureScreen('issue0');
      }, undefined, 1000);

      casper.waitForSelector('.notExisting1', function() {
        test.assertExists('.notExisting1', 'THIS SHOULD NOT HAPPEN');
        this.captureScreen('issue1');
      }, undefined, 1000);
      casper.run(function() { test.done(); });
    });

Run it with `casperjs  test nastyTest.js niceTest.js` (tested with `1.1.3`)

As you can see `niceTest.js` should always pass, but if you run it several times you will
see it sometimes fails, with

    In niceTest.js
      Nice test
        uncaughtError: ".notExisting1" still did not exist in 1000ms

What I think is happening is:

The first test hits the `waitFor.timeout`, which ends up calling [`ErrorHandlerAndDone`](https://github.com/casperjs/casperjs/blob/master/modules/tester.js#L183)

This calls `this.done`, which sets [`this.running=false`](https://github.com/casperjs/casperjs/blob/master/modules/tester.js#L1236)

This then allows the interval from `runSuites` to move to the next test.

Meanwhile, the `Casper` instance is running `checkStep`. If this fires at the right
moment then it will see that the previous step has finished, and so launch the new
one. This runs the `waitFor` which creates another interval.

If that happens then at some point `waitFor` will hit the timeout, which will
then trigger the tester to report the second test as failed.

The problem is that the two test runs aren't completely isolated. One fix
that seems to work for me is to add `casper.bypass` to `tester.done` so that
it clears the remaining steps before letting the next test execute.

Another options would be for `casper.start` to cancel any intervals in
`this.waiters` so that they won't interfere with the next use of casper.
This is because this is something that could happen without the tester
involved, so would warrant a new solution.

Thanks,

James

**Comment by james-w** [comment]
Btw, 10 runs of the test scripts are usually enough to trigger the problem at least once.

**Comment by james-w** [comment]
also @javsalgar is the one that found and isolated the problem.

**Comment by kensoh** [comment]
Hi @james-w thanks for raising this, and providing deep details. I'm a CasperJS user for couple of years now browsing through open issues to see if there is something I can help with or learn more. This is a use case which I can't contribute but wanted to watch and follow to see how it develops.

**Comment by kensoh** [comment]
Bumping up the thread and also copying maintainer @jefleponot

**Comment by kensoh** [comment]
Bumping -

**Comment by kensoh** [comment]
Archiving issue using ARCHIVE label as CasperJS, like PhantomJS, is not actively maintained for now.
